include/cadmium/celldevs/cell/cell.hpp

```diff
@@ -0,0 +1,224 @@
+         */
+        explicit cell_state_message(C cell_id_in, S state_in): cell_id(cell_id_in), state(state_in) {}
+
+        /**
+         * Operator overloading function for printing cell state messages
+         * @tparam C the used for representing a cell ID.
+         * @tparam S the type used for representing a cell state.
+         * @param os Operating System's string stream
+         * @param msg cell state message
+         * @return Operating System's string stream containing the stringified cell state
+         */
+        friend std::ostream &operator << (std::ostream &os, const cell_state_message<C, S> &msg) {
+            os << msg.cell_id << " ; " << msg.state;
+            return os;
+        }
+    };
+
+
+    /**
+     * @brief input/output port structure for cells.
+     * @tparam C the used for representing a cell ID.
+     * @tparam S the type used for representing a cell state.
+     */
+    template <typename C, typename S>
+    struct cell_ports_def{
+        struct cell_in: public cadmium::in_port<cell_state_message<C, S>> {};
+        struct cell_out : public cadmium::out_port<cell_state_message<C, S>> {};
+    };
+
+    /**
+     * @brief Abstract DEVS atomic model for defining cells in Cell-DEVS scenarios.
+     * @tparam T the type used for representing time in a simulation.
+     * @tparam C the used for representing a cell ID.
+     * @tparam S the type used for representing a cell state.
+     * @tparam V the type used for representing a neighboring cell's vicinity. By default, it is set to integer.
+     * @tparam C_HASH the hash function used for creating unordered maps with cell IDs as keys.
+     *                By default, it uses the one defined in the standard library
+     */
+    template <typename T, typename C, typename S, typename V=int, typename C_HASH=std::hash<C>>
+    class cell {
+    public:
+
+        using input_ports = std::tuple<typename cell_ports_def<C, S>::cell_in>;
+        using output_ports = std::tuple<typename cell_ports_def<C, S>::cell_out>;
+
+        using NV = std::unordered_map<C, V, C_HASH>;
+        using NS = std::unordered_map<C, S, C_HASH>;
+
+
+        C cell_id;                          /// Cell ID
+        std::vector<C> neighbors;           /// Neighboring cells' IDs
+        T clock;                            /// simulation clock
+        T next_internal;                    /// Time remaining until next internal state transition
+        delayer<T, S> *buffer;              /// output message buffer
+
+        struct state_type {
+            S current_state;                /// Cell's internal state
+            NV neighbors_vicinity;          /// Neighboring cell' vicinity type
+            NS neighbors_state;             /// neighboring cell' public state
+        };
+        state_type state;
+
+        /// A default constructor is required for compiling issues. However, it is not valid and always throws exception
+        cell(){ throw std::exception(); }
```

Comment:
Why? Can you explain a little? It doesn't sound right.

Reply:
I expect at least a Cell ID. Furthermore, a cell needs an initial state, its neighbors (and, if needed, their vicinity), and the type of delay to be used (either inertial or transport). However, I think that Cadmium needs the models to have a valid empty constructor to compile. That is why I just throw an exception.

Reply:
Ideally you should look into why a constructor is required, and if it makes no sense, remove the requirement. However, even if we do not adjust the requirements, at least put a more specific exceptiontype and a message to know where that exception came from when it is raised.